packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts

```diff
@@ -436,6 +448,43 @@ function isConditionalTest(node: TSESTree.Node): boolean {
   return false;
 }
 
+function isMakeBoolean(node: TSESTree.Node): boolean {
+  const parents = new Set<TSESTree.Node | null>([node]);
+  let current = node.parent;
+  while (current) {
+    parents.add(current);
+
+    if (
+      (current.type === AST_NODE_TYPES.UnaryExpression &&
+        current.operator === '!') ||
+      (current.type === AST_NODE_TYPES.CallExpression &&
+        current.callee.type === AST_NODE_TYPES.Identifier &&
+        // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+        current.callee.name === 'Boolean')
+    ) {
+      return true;
+    }
+
+    if (
+      [
+        AST_NODE_TYPES.ArrowFunctionExpression,
+        AST_NODE_TYPES.FunctionExpression,
+      ].includes(current.type)
+    ) {
+      /**
+       * This is a weird situation like:
+       * `if (() => a || b) {}`
+       * `if (function () { return a || b }) {}`
+       */
+      return false;
```

Comment:
[Testing] Good to see that the cases tripping this up are known 😄.  

Also to test are other unusual things in `if`s:

* `class` expressions
* `new` expressions
* Function decorators
* ...I'm drawing a blank on more spooky things at the moment, but there's a good chance more exist.

Reply:
1. Because I was copying the isConditionalTest function, I forgot to change the example case. In practice, an example would be `Boolean(()=> a || b)`. 
Therefore, it seems that the situations you mentioned should actually apply to the isConditionalTest function as well. Should the work be done in this pr?

2. Actually, I did not exactly understand the three cases you mentioned. Could you please give an example if possible?

3. The additional case I was considering is an object. Should cases like `Boolean({test:a||b})` also be included?